### DISCUSSION TITLE
How to delete default nvim parser? Getting 'query: invalid node type' error

### DISCUSSION THREAD

**Discussion by Ashe** [nvim-treesitter/nvim-treesitter#5289] [Q&A] [answered]
Hello, for the longest time I've been dealing with tedious issues with treesitter. It sometimes goes away, sometimes doesn't, but I've had enough. I saw on the FAQ that this is a common error caused by multiple parsers. I ran `:echo nvim_get_runtime_file('parser', v:true)` and got this:

```
['/nix/store/avmr95wgfjyzi6n1axbjx9x9952s1d01-neovim-unwrapped-0.9.1/lib/nvim/parser', '/home/ashe/.local/share/nvim-treesitter/parsers/parser']  
```
I can't figure out for the life of me how to disable the standard nvim parser, or any sort of acknowledgement online that this is a feature and how to get rid of it. I hope someone knows? 

Thank you!

**Comment by clason** [comment]
You can't (or rather, shouldn't). You need to make sure that the parser directory has the same precedence over the bundled directory that the query directory has over the one in $VIMRUNTIME.

How to do that with Nix, I cannot tell you, sorry.

**Reply by Ashe** [reply, marked as answer]
Thank you for your reply, I'm trying to use the package `pkgs.vimPlugins.nvim-treesitter.withAllGrammers`. The command that you suggest to run now fills my screen with parsers, but at the same time I'm not seeing the errors anymore, so maybe this is how we should be setting this up on nix.

Thank you for letting me know this wasn't really a config issue and more of a nix thing :)

**Comment by ecerulm** [comment]
one way is to manipulate the `vim.opt.rtp` (runtimepath)

for example, I originally have multiple paths for `parser` like you  


```vim
:echo nvim_get_runtime_file('parser', v:true)
['/Users/rubelagu/.local/share/nvim/lazy/nvim-treesitter/parser', '/usr/local/Cellar/neovim/0.9.5/lib/nvim/parser']
```

I can remove the `/usr/local/Cellar/neovim/0.9.5/lib/nvim` from `runtimepath`(`:h 'runtimepath`) using `vim.opt.rtp`.  Beware if you are using lazy.nvim as plugin manager since it will reset the `runtimepath` unless you set the `performance.rtp.reset=false`

```lua
vim.opt.rtp:remove("/usr/local/Cellar/neovim/0.9.5/lib/nvim") -- this is the important bit 
require("lazy").setup("ecerulm.plugins", {
  -- https://github.com/folke/lazy.nvim
  performance = {
    rtp = {
      reset = false -- I need this because otherwise lazy.nvim reset the runtimepath
    },
  },
	change_detection = {
		enabled = true, -- automatically check for config file changes and reload the ui
		notify = false, -- turn off notifications whenever plugin changes are made
	},
})

```

Alternatively you can put the 

```lua
vim.opt.rtp:remove("/usr/local/Cellar/neovim/0.9.5/lib/nvim") 
```

after the lazy.nvim and that should work too.

**Reply by clason** [reply]
No, that is very dangerous advice. _Never_ mess with the bundled runtime! Instead you should make sure that nvim-treesitter takes precedence on your runtimepath; this should normally be the case, and if Lazy messes this up, it should be fixed. (DO NOT lazy-load nvim-treesitter; that is not supported!)

**Reply by ecerulm** [reply]
The only thing that goes in  `/usr/local/Cellar/neovim/0.9.5/lib/nvim` is the `parser`

```
find /usr/local/Cellar/neovim/0.9.5/lib/nvim
/usr/local/Cellar/neovim/0.9.5/lib/nvim
/usr/local/Cellar/neovim/0.9.5/lib/nvim/parser
/usr/local/Cellar/neovim/0.9.5/lib/nvim/parser/c.so
/usr/local/Cellar/neovim/0.9.5/lib/nvim/parser/vim.so
/usr/local/Cellar/neovim/0.9.5/lib/nvim/parser/vimdoc.so
/usr/local/Cellar/neovim/0.9.5/lib/nvim/parser/lua.so
/usr/local/Cellar/neovim/0.9.5/lib/nvim/parser/query.so
``` 

But you are right in that by disabling the lazy loading 

```
{
	"nvim-treesitter/nvim-treesitter",
         lazy = false,
```

it works without having to remove the  that directory from the `runtimepath`.

**Reply by clason** [reply]
> The only thing that goes in  /usr/local/Cellar/neovim/0.9.5/lib/nvim is the parser

Yes, and they are there for a reason. Removing them (without installing a compatible(!) version externally) _will_ break Neovim.

**Comment by JavonneM** [comment]
I had an issue with `invalid node type method_element` 
All I had to do was a `:TSUpdate` to solve my issue. I did however also change my treesitter `lazy = false`

**Reply by curiousponge** [reply]
Thank you! I was experiencing that error on snacks.picker sometimes, so I was looking for a solution. Yours is a much cleaner and simpler solution... and it worked :)
